File: pipeline/post_production.py
import json
import os
from pathlib import Path
import tempfile
import time
import re

import boto3

import ffmpeg


def main(args):
    # Get the service client.
    s3_client = boto3.client('s3')

    key = args.get('key', args.get('object_name', ''))

    # parse the key
    choir_id, song_id, def_id = Path(key).stem.split('-')[0].split('+')

    src_bucket = os.environ['SRC_BUCKET']
    dst_bucket = os.environ['DEST_BUCKET']
    misc_bucket = os.environ['MISC_BUCKET']
    definition_bucket = os.environ['DEFINITION_BUCKET']

    # Download the definition file for this job
    definition_key = f'{choir_id}+{song_id}+{def_id}.json'
    definition_object = s3_client.get_object(
        Bucket=definition_bucket,
        Key=definition_key,
    )
    definition = json.load(definition_object['Body'])
    output_spec = definition['output']

    ###
    # Combine video and audio
    ###

    # Create a temp dir for our files to use
    with tempfile.TemporaryDirectory() as tmpdir:

        print("Doing first pass")
        src_url = s3_client.generate_presigned_url(
            ClientMethod='get_object',
            Params={
                'Bucket': src_bucket,
                'Key': key
            }
        )
        stream = ffmpeg.input(src_url,
                              seekable=0)
        audio = stream.audio
        audio = audio.filter('volumedetect')
        pipeline = ffmpeg.output(audio,
                                 "-",
                                 format='null')

        cmd = pipeline.compile()
        print("ffmpeg command to run: ", cmd)

        stdout, stderr = pipeline.run(capture_stdout=True,
                                      capture_stderr=True)
        output = stdout + stderr
        output_lines = [line.strip() for line in output.decode().split('\n')]

        # mute = False

        # Volume detect
        vol_threshold = int(args.get('vol_threshold', 22))
        vol_pct = float(args.get('vol_pct', 0.05))

        total_samples = 0
        high_samples = 0
        max_volume = 0
        hist_re = re.compile(r'histogram_(\d+)db: (\d+)')
        maxvol_re = re.compile(r'max_volume: (-?\d+\.?\d*) dB')
        for line in output_lines:
            # Search for histogram
            mo = hist_re.search(line)
            if mo:
                level, samples = mo.groups()
                total_samples += int(samples)
                if int(level) < vol_threshold:
                    high_samples += int(samples)

            # Search for max volume
            mo = maxvol_re.search(line)
            if mo:
                max_volume = float(mo.groups()[0])

        if high_samples / total_samples < vol_pct:
            print(f"Input volume is so low, we are muting it "
                  f"{high_samples/total_samples:.2f} above {vol_threshold}")
            # mute = True

        target_peak = 0
        volume_gain = target_peak - max_volume
        volume_gain = f"{volume_gain:.2f} dB"

        # Second pass, apply normalisation
        print("Doing second pass loudnorm")
        stream = ffmpeg.input(src_url,
                              seekable=0)

        video = stream.video
        audio = stream.audio

        # Pad the video to final size, place video in center
        output_width, output_height = output_spec['size']
        video = video.filter('pad',
                             x=-1,
                             y=-1,
                             width=output_width,
                             height=output_height)

        # Overlay the watermark if present
        watermark_file = output_spec.get('watermark')
        if watermark_file:
            watermark_url = s3_client.generate_presigned_url(
                ClientMethod='get_object',
                Params={
                    'Bucket': misc_bucket,
                    'Key': watermark_file
                }
            )
            watermark = ffmpeg.input(watermark_url,
                                     seekable=0)
            video = video.overlay(watermark,
                                  x='W-w-20',
                                  y='H-h-20')

        print("Volume gain to apply:", volume_gain)
        audio = audio.filter('volume',
                             volume_gain)

        # Add in audio compression
#        audio = audio.filter('acompressor')

        # Add reverb in if present
        reverb_type = output_spec.get('reverb_type')
        if reverb_type:
            reverb_url = s3_client.generate_presigned_url(
                ClientMethod='get_object',
                Params={
                    'Bucket': misc_bucket,
                    'Key': f'{reverb_type}.wav'
                }
            )
            reverb_pct = float(output_spec.get('reverb', 0.1))
            if reverb_pct > 0:
                reverb_part = ffmpeg.input(reverb_url,
                                           seekable=0)
                split_audio = audio.filter_multi_output('asplit')
                reverb = ffmpeg.filter([split_audio[1], reverb_part],
                                       'afir',
                                       dry=10, wet=10)
                audio = ffmpeg.filter([split_audio[0], reverb],
                                      'amix',
                                      dropout_transition=180,
                                      weights=f'{1-reverb_pct} {reverb_pct}')

        # Output
        output_key = f'{choir_id}+{song_id}+{def_id}-final.mp4'
        output_path = str(Path(tmpdir, output_key))

        kwargs = {}
        if 'duration' in args:
            kwargs['t'] = int(args['duration'])

        if 'loglevel' in args:
            kwargs['v'] = args['loglevel']

        pipeline = ffmpeg.output(audio,
                                 video,
                                 output_path,
                                 pix_fmt='yuv420p',
                                 vcodec='libx264',
                                 preset='veryfast',
                                 movflags='+faststart',
                                 **kwargs
                                 )
        cmd = pipeline.compile()
        print("ffmpeg command to run: ", cmd)
        t1 = time.time()
        pipeline.run()
        t2 = time.time()

        # write the output file to S3
        s3_client.upload_file(output_path, dst_bucket, output_key)

        ret = {'dst_key': output_key,
               'def_id': def_id,
               'render_time': int(t2 - t1),
               'choir_id': choir_id,
               'song_id': song_id,
               'status': 'done'}

        return ret

File: pipeline/post_production.js
// environment variables
const LOCAL_MODE = !!process.env.LOCAL_MODE
const MISC_BUCKET = process.env.MISC_BUCKET
const DEST_BUCKET = process.env.DEST_BUCKET
const DEFINITION_BUCKET = process.env.DEFINITION_BUCKET
const LOCAL_BUCKETS_PATH = '../buckets'

// node modules
const fs = require('fs')
const path = require('path')
const aws = require('aws-sdk')
const tmpdir = require('tmpdir')
const ffmpeg = require('fluent-ffmpeg')
const run = require('ffmpegrunner').run

// aws objects
const S3 = new aws.S3()

// break file name into constituent bits
const parseKey = (key) => {
  // split the incoming key into bits
  const parsed = path.parse(key).name.split('+')
  const pk = {
    choirId: parsed[0],
    songId: parsed[1]
  }
  const bits = parsed[2].split('-')
  pk.defId = bits[0]
  return pk
}

// build the filter graph for processing audio & video
const buildComplexFilter = (outputWidth, outputHeight, reverbLevel) => {
  // there are three inputs
  // 0 - the input video & audio
  // 1 - the watermark
  // 2 - the reverb impulse response wav
  // complex filter
  const filters = []
  let f

  // video pipeline
  f = {
    inputs: '0:v',
    filter: 'pad',
    options: {
      x: -1,
      y: -1,
      width: outputWidth,
      height: outputHeight
    },
    outputs: 'v1'
  }
  filters.push(f)
  f = {
    inputs: ['v1', '1'],
    filter: 'overlay',
    options: {
      x: 'W-w-20',
      y: 'H-h-20'
    },
    outputs: 'video'
  }
  filters.push(f)

  // audio pipeline
  f = {
    inputs: '0:a', // audio on the main video
    filter: 'asplit',
    options: { },
    outputs: ['a1', 'a2']
  }
  filters.push(f)
  f = {
    inputs: ['a1', '2'], // audio+ reverb impulse response
    filter: 'afir',
    options: {
      dry: 10,
      wet: 10
    },
    outputs: ['reverb']
  }
  filters.push(f)
  f = {
    inputs: ['a2', 'reverb'], // reverb + original audio split
    filter: 'amix',
    options: {
      inputs: 2,
      dropout_transition: 180,
      weights: `${1 - reverbLevel} ${reverbLevel}`
    },
    outputs: ['audiomix']
  }
  filters.push(f)

  // return the list of filters and the main audio & video outputs
  return {
    filters: filters,
    outputs: ['video', 'audiomix']
  }
}

// main
const main = async (event, context) => {
  console.log('post_production')

  let key, bucket, geturl, definition, watermarkurl, reverburl

  // when running locally
  if (LOCAL_MODE) {
    // expect 'event' to be a flat object
    key = event.key
    bucket = event.bucket
  } else {
    // event is an AWS event
    key = unescape(event.Records[0].s3.object.key)
    bucket = event.Records[0].s3.bucket.name
  }

  // parse the key
  const pk = parseKey(key)

  // load the definition file
  const definitionKey = `${pk.choirId}+${pk.songId}+${pk.defId}.json`
  if (LOCAL_MODE) {
    const p = path.join(LOCAL_BUCKETS_PATH, DEFINITION_BUCKET, definitionKey)
    definition = fs.readFileSync(p, { encoding: 'utf8' })
  } else {
    definition = await S3.getObject({ Bucket: DEFINITION_BUCKET, Key: definitionKey }).promise()
    definition = definition.Body
  }
  definition = JSON.parse(definition)

  // calculate input urls for video file and watermark image
  const reverbFile = `${definition.output.reverb_type}.wav`
  if (LOCAL_MODE) {
    geturl = path.join(LOCAL_BUCKETS_PATH, bucket, key)
    watermarkurl = path.join(LOCAL_BUCKETS_PATH, MISC_BUCKET, definition.output.watermark)
    reverburl = path.join(LOCAL_BUCKETS_PATH, MISC_BUCKET, reverbFile)
  } else {
    geturl = await S3.getSignedUrlPromise('getObject', { Bucket: bucket, Key: key })
    watermarkurl = await S3.getSignedUrlPromise('getObject', { Bucket: MISC_BUCKET, Key: definition.output.watermark })
    reverburl = await S3.getSignedUrlPromise('getObject', { Bucket: MISC_BUCKET, Key: reverbFile })
  }
  console.log(geturl, watermarkurl, reverburl)

  // build ffmpeg command
  const command = ffmpeg()
    .addInput(geturl)
    .inputOptions(['-seekable 0'])
    .addInput(watermarkurl)
    .inputOptions(['-seekable 0'])
    .addInput(reverburl)
    .inputOptions(['-seekable 0'])

  // build the video & audio pipelines
  const complexFilter = buildComplexFilter(
    definition.output.size[0], // width of output video
    definition.output.size[1], // height of output video
    definition.output.reverb)
  command.complexFilter(complexFilter.filters, complexFilter.outputs)

  // create temporary directory
  const tmppath = tmpdir.createTmpDirectory()
  const outputKey = `${pk.choirId}+${pk.songId}+${pk.defId}-final.mp4`
  const outpath = path.join(tmppath, outputKey)

  // set output parameters
  command
    .output(outpath)
    .outputFormat('mp4') // mp4 container
    .outputOptions([
      '-pix_fmt yuv420p',
      '-vcodec libx264', // h.264 video
      '-preset veryfast', // fast encoding
      '-movflags +faststart']) // put meta data at the start of the file

  await run(command, true)

  // copy the temporary file to output bucket
  if (LOCAL_MODE) {
    const desturl = path.join(LOCAL_BUCKETS_PATH, DEST_BUCKET, outputKey)
    fs.copyFileSync(outpath, desturl)
  } else {
    // upload to S3
    await S3.putObject({
      Bucket: DEST_BUCKET,
      Key: outputKey,
      Body: fs.createReadStream(outpath)
    }).promise()
  }
  tmpdir.removeTmpDirectory(tmppath)
  return { ok: true }
}

exports.main = main
